src/Hazelcast.Net.Tests/Serialization/Compact/GenericRecordTests.cs

```diff
@@ -0,0 +1,406 @@
+﻿// Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+// http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Threading.Tasks;
+using Hazelcast.Models;
+using Hazelcast.Serialization;
+using Hazelcast.Serialization.Compact;
+using Hazelcast.Testing;
+using Moq;
+using NUnit.Framework;
+
+namespace Hazelcast.Tests.Serialization.Compact;
+
+[TestFixture]
+public class GenericRecordTests : SingleMemberClientRemoteTestBase
+{
+    [Test]
+    public void TestLocal()
+    {
+        var rec0 = GenericRecordBuilder.Compact("rec0")
+            .SetBoolean("bool", true)
+            .SetInt32("int", 123)
+            .SetString("string", "hello")
+            .Build();
+
+        Assert.That(rec0.GetBoolean("bool"), Is.True);
+        Assert.That(rec0.GetInt32("int"), Is.EqualTo(123));
+        Assert.That(rec0.GetString("string"), Is.EqualTo("hello"));
+
+        Assert.Throws<SerializationException>(() => rec0.GetInt32("meh"));
+        Assert.Throws<SerializationException>(() => rec0.GetString("int"));
+
+        var now = (HLocalDate) DateTime.Now;
+
+        var rec1 = GenericRecordBuilder.Compact("rec1")
+            .SetDate("date", now)
+            .SetGenericRecord("rec", rec0)
+            .Build();
+
+        Assert.That(rec1.GetDate("date"), Is.EqualTo(now));
+
+        var rec1A = rec1.GetGenericRecord("rec");
+        Assert.That(rec1A, Is.Not.Null);
+        if (rec1A is null) throw new Exception(); // we *know* it's not null but R# does not
+
+        Assert.That(rec1A.GetBoolean("bool"), Is.True);
+        Assert.That(rec1A.GetInt32("int"), Is.EqualTo(123));
+        Assert.That(rec1A.GetString("string"), Is.EqualTo("hello"));
+
+        var rec2 = GenericRecordBuilder.Compact("rec2")
+            .SetInt32("int", 123)
+            .Build();
+
+        Assert.Throws<SerializationException>(() => rec2.NewBuilder().Build());
+
+        var unused0 = rec2.NewBuilderWithClone().Build();
+        var unused1 = rec2.NewBuilder().SetInt32("int", 567).Build();
+
+        Assert.Throws<SerializationException>(() => rec2.NewBuilderWithClone().SetInt32("other", 567).Build());
+
+        Assert.Throws<ArgumentException>(() => 
```

Comment:
So, this shows us user can't set a generic record which is not build by the builder, right?

Reply:
Correct. In fact it has to be a compact generic record (in Java there are also portable generic records).